### ISSUE TITLE
Asterisk Java road map proposal

### ISSUE THREAD

**Issue by piotrooo** [asterisk-java/asterisk-java#405]
Hello :wave:

Dear @asterisk-java/contributors, @asterisk-java/owners and everyone who want to support us!

I would like to invite all of you to this conversation.
These are the next steps of making the asterisk-java project great again.

---

### :exclamation: Disclaimer 
This is only a *proof-of-concept* and should be treated as a *proposal* and a **base for the discussion**.

<sub>I hope responses will be faster.</sub>

---

> :bangbang:
> If you don't agree with any changes or proposals, **let me know** by using the reaction button, or more helpfully, by giving ***constructive feedback*** on what is wrong and how it should be handled.

---

I've diagnosed a couple of problematic areas.

### :boom: Problems:
* Right now, releases are done by some weird Dart release scripts, which are neither Java nor GitHub friendly. 
* We don't have any changelog. Nowadays, this is bad practice and very uninformative.
* In addition, a lot of changes are made without any issue or even a short description of the reason for changes.

### :bulb: Solutions:
* Do release using GitHub Releases connected with GitHub Actions and Maven Release Plugin (or another way, but using Java).
* GitHub Releases helps us with the changelog.
* All changes must have an appropriate issue or pull request with a short description. Master (maybe main?) should be blocked for direct pushes. Every change should be done via pull requests (for everyone!) and reviewed.

### :boom: Problems:
* Right now everything is mixed into one module. If you want to run only fastagi or manager - you can't.
* There aren't boundaries between modules.

### :bulb: Solutions:
I suggest the following project structure:

```
asterisk-java-parent        (POM)
├── asterisk-java-agi       (POM)
│   ├── asterisk-java-agi-asyncagi  (JAR)    dependencies: asterisk-java-agi-command, asterisk-java-agi-core, asterisk-java-ami-manager
│   ├── asterisk-java-agi-command   (JAR)
│   ├── asterisk-java-agi-core      (JAR)
│   └── asterisk-java-agi-fastagi   (JAR)    dependencies: asterisk-java-agi-core, asterisk-java-agi-command
├── asterisk-java-ami       (POM)
│   ├── asterisk-java-ami-action    (JAR)
│   ├── asterisk-java-ami-event     (JAR)
│   └── asterisk-java-ami-manager   (JAR)   dependencies: asterisk-java-ami-action, asterisk-java-ami-event
├── asterisk-java-common    (JAR)
├── asterisk-java-live      (JAR)           dependencies: asterisk-java-ami-manager
└── asterisk-java-pbx       (JAR)           dependencies: asterisk-java-agi-fastagi, asterisk-java-ami-action, asterisk-java-ami-event
```

Why are modules good? 
* They give an overview of how asterisk-java and Asterisk itself are designed. They also provide boundaries of responsibility.
* Prevent mixing objects (exceptions, models, etc.) between modules.
* Gives possibility to extend in the future (we can easily add e.g. ARI module).

### :boom: Problems:
Currently the project doesn't have clearly defined Asterisk and Java supported versions.

#### Java
* All LTS versions shlould be supported and tested (8, 11, 17).
* If any Java version becomes EOL, we should drop support in the next major release (for Java 8 it's 2025-03-31).

#### Asterisk
For Asterisk versions 16, 18 and 20 the major should be 4.x.x. When 16 reaches its EOL in 2023-10-09, we should change
the major to 5.x.x. We should also change the major version when a new Asterisk LTS version is released.

##### asterisk-java project release cycle:

| Asterisk Java | Asterisk   | Java          | EOL        | Reason                                                       |
|---------------|------------|---------------|------------|--------------------------------------------------------------|
| 4.x.x         | 16, 18, 20 | 8, 11, 17     | 2023-10-09 | Java 21 released<br/>Asterisk 16 EOL                         |
| 5.x.x         | 18, 20     | 8, 11, 17, 21 | 2024-10-19 | Asterisk 22 released                                         |
| 6.x.x         | 18, 20, 22 | 8, 11, 17, 21 | 2025-03-31 | Java 8 EOL                                                   |
| 7.x.x         | 18, 20, 22 | 11, 17, 21    | 2025-10-20 | Asterisk 18 EOL                                              |
| 8.x.x         | 20, 22     | 11, 17, 21    | 2026-10-19 | Java 11 EOL<br/>Java 27(?) released<br/>Asterisk 24 released |

#### :exclamation: Disclaimer
With this release process I have a lot of confusion. I'm not 100% sure it's a good way to do it.

### :boom: Problems:
* Without labels, we cannot make any automation.
* Also issues and pull requests are uncategorized.

### :bulb: Solutions:
Create labels and use it.

Connected with particular Asterisk version.
```
asterisk/16
asterisk/18
asterisk/20
...
```

Type of issue or pull request.
```
kind/bug
kind/breaking
kind/dependencies
kind/documentation
kind/enhancement
kind/feature
kind/question
kind/refactor
kind/wontfix
```

Additional information for issue or pull request.
```
meta/skip-changelog
```

Connected for particular module.
```
module/agi
module/ami
module/live
module/pbx
```

### :boom: Problem:
There is a bunch of the stale branches.

### :bulb: Solution:
Remove all of them, and leave only master.

### :boom: Problem:
Right now, there aren't any integration tests. We are not sure that our lib working with supported Asterisk versions.

### :bulb: Solution:
We should running integration test for each supported Asterisk version. To achieve this, we can use Testcontainers.

Here is a proof-of-concept of super easy and a little bit dumb example, but this prove it's possible to run integration tests automatically.

https://github.com/piotrooo/asterisk-java/commit/dbefe8ba44260a72d7090caf7b46e4709343c63c

Tests should be run for all supported Asterisk versions.

### :boom: Problem:
Web page is very outdated.

### :bulb: Solution:
* Create new web page with introduction, modules overview, documentation, some explanations.
* We should use some static site generator like [MkDocs](https://www.mkdocs.org/) or [AsciiDoc](https://asciidoc.org/) or some other.

### :boom: Problem:
Last but not least, asterisk-java doesn't have own logo :sob:

### :bulb: Solution:
I've asked my good friend @tomekbbk, who is also a graphic and UX designer in company where I've working, to create logo.
This is what he designed for us:

![Logo_b_textmdpi](https://user-images.githubusercontent.com/2005054/194570978-a0a9178b-4655-49ea-a92a-7de0840e7029.png)

![Sygnetmdpi](https://user-images.githubusercontent.com/2005054/194570998-367b1703-ba58-48f1-a581-d674380b9c69.png)

BTW. Connect Java coffee cup with Asterisk logo = FTW :star_struck:

---

###### In my opinion

1. Create labels for issues and pull requests and start to use them
2. Cleanup release cycle - use maven plugin and generate change log (GitHub Actions and GitHub Releases)
3. Block master for direct pushes
4. Cleanup stale branches
5. Try to address and fix at least couple of issues
6. Modularize project
7. Add integration tests
8. Web page

_After that changes, version **4.0.0** should be  #released!_

---

**Comment by bsutton** [comment]
Wow and ok.

Like the logo.

My biggest concern is that it is a big list of things and big lists can
stall an open source project.

**Comment by piotrooo** [comment]
> My biggest concern is that it is a big list of things and big lists can stall an open source project.

Indeed, list of things are really big. Doing small steps it's the key.

**Comment by rlsutton1** [comment]
There is a git hub action for releases, the dart code only updates the version number and creates a git tag which triggers the git hub action.

**Comment by piotrooo** [comment]
> There is a git hub action for releases, the dart code only updates the version number and creates a git tag which triggers the git hub action.

Yes, I saw it.
Right now this process looks. Someone runs a Dart script, which:
* create tag
* make some changes at local code base 
* push tag

And _**after**_ that GitHub Actions publish an artifact.
If someone is a Dart noob (as I'm), don't know how to release project version. I don't know how to setup environment. 

I suggest reverse this process and make all things in GitHub Releases. Which creates tag, generates changelog, produces and publishes artifact, in the future do something with docs. What do you think?

BTW. @rlsutton1 what do you think about rest of proposal. You and @bsutton are now active maintainers of this project, so I would glad about you opinion :smile:
